### PR TITLE
Fix Either to run transform on the first param that pass validation

### LINKED ISSUE #11830: [BUG] TypeError when customizing tools icon using PIL Image
### Software versions

```
Python version      :  3.9.6 | packaged by conda-forge | (default, Jul  6 2021, 08:53:59)
IPython version     :  7.29.0
Tornado version     :  6.1
Bokeh version       :  3.0.0dev1+21.g51a06bb37.dirty
BokehJS static path :  /home/laptop/projects/bokeh/bokeh/server/static
node.js version     :  v16.6.1
npm version         :  7.20.3
```

### Browser name and version

_No response_

### Jupyter notebook / Jupyter Lab version

notebook: 6.4.5

### Expected behavior

Show the figure, along with the added tool, and render the custom icons instead of the default tool icon.

### Observed behavior

Code doesn't run, and raises a `TypeError: Object of type Image is not JSON serializable` more details in the traceback.

### Example code

```Python
from bokeh.plotting import figure, show
from bokeh.models.tools import BoxSelectTool
from PIL import Image
from matplotlib import cm
import numpy as np

myarray = np.array([
    [1, 1, 1, 1], 
    [1, 1, 1, 1], 
    [0, 0, 0, 0],])
im = Image.fromarray(np.uint8(cm.gist_earth(myarray)*255))

x = [1, 2, 3, 4, 5]
y = [6, 7, 2, 4, 5]

# create a new figure
figure = figure(title="",tools="")
figure.add_tools(BoxSelectTool(description="Graph selector", icon=im))

# add a line renderer with legend and line thickness
figure.line(x, y, legend_label="Temp.", line_width=2)

# show the results
show(p)
```

### Stack traceback or browser console output

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_962/3797676593.py in <module>
     11 
     12 # show the results
---> 13 show(p)

~/projects/bokeh/bokeh/io/showing.py in show(obj, browser, new, notebook_handle, notebook_url, **kwargs)
    142 
    143     if isinstance(obj, LayoutDOM):
--> 144         return _show_with_state(obj, state, browser, new, notebook_handle=notebook_handle)
    145 
    146     def is_application(obj: Any) -> TypeGuard[Application]:

~/projects/bokeh/bokeh/io/showing.py in _show_with_state(obj, state, browser, new, notebook_handle)
    193 
    194     if state.file or not shown:
--> 195         _show_file_with_state(obj, state, new, controller)
    196 
    197     return comms_handle

~/projects/bokeh/bokeh/io/showing.py in _show_file_with_state(obj, state, new, controller)
    174 
    175     '''
--> 176     filename = save(obj, state=state)
    177     controller.open("file://" + filename, new=NEW_PARAM[new])
    178 

~/projects/bokeh/bokeh/io/saving.py in save(obj, filename, resources, title, template, state)
     96 
     97     filename, resources, title = _get_save_args(state, filename, resources, title)
---> 98     _save_helper(obj, filename, resources, title, template, theme)
     99     return abspath(filename)
    100 

~/projects/bokeh/bokeh/io/saving.py in _save_helper(obj, filename, resources, title, template, theme)
    162     '''
    163     from ..embed import file_html
--> 164     html = file_html(obj, resources, title=title, template=template or FILE, theme=theme)
    165 
    166     with open(filename, mode="w", encoding="utf-8") as f:

~/projects/bokeh/bokeh/embed/standalone.py in file_html(models, resources, title, template, template_variables, theme, suppress_callback_warning, _always_new)
    332 
    333     with OutputDocumentFor(models_seq, apply_theme=theme, always_new=_always_new) as doc:
--> 334         (docs_json, render_items) = standalone_docs_json_and_render_items(models_seq, suppress_callback_warning=suppress_callback_warning)
    335         title = _title_from_models(models_seq, title)
    336         bundle = bundle_for_objs_and_resources([doc], resources)

~/projects/bokeh/bokeh/embed/util.py in standalone_docs_json_and_render_items(models, suppress_callback_warning)
    308     docs_json: Dict[ID, DocJson] = {}
    309     for doc, (docid, _) in docs.items():
--> 310         docs_json[docid] = doc.to_json()
    311 
    312     render_items: List[RenderItem] = []

~/projects/bokeh/bokeh/document/document.py in to_json(self)
    734         # this is a total hack to go via a string, needed because
    735         # our BokehJSONEncoder goes straight to a string.
--> 736         doc_json = self.to_json_string()
    737         return loads(doc_json)
    738 

~/projects/bokeh/bokeh/document/document.py in to_json_string(self, indent)
    766         )
    767 
--> 768         return serialize_json(json, indent=indent)
    769 
    770     def unhold(self) -> None:

~/projects/bokeh/bokeh/core/json_encoder.py in serialize_json(obj, pretty, indent, **kwargs)
    165         indent = 2
    166 
--> 167     return json.dumps(obj, cls=BokehJSONEncoder, allow_nan=False, indent=indent, separators=separators, sort_keys=True, **kwargs)
    168 
    169 

~/miniconda3/envs/bkdev/lib/python3.9/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    232     if cls is None:
    233         cls = JSONEncoder
--> 234     return cls(
    235         skipkeys=skipkeys, ensure_ascii=ensure_ascii,
    236         check_circular=check_circular, allow_nan=allow_nan, indent=indent,

~/miniconda3/envs/bkdev/lib/python3.9/json/encoder.py in encode(self, o)
    197         # exceptions aren't as detailed.  The list call should be roughly
    198         # equivalent to the PySequence_Fast that ''.join() would do.
--> 199         chunks = self.iterencode(o, _one_shot=True)
    200         if not isinstance(chunks, (list, tuple)):
    201             chunks = list(chunks)

~/miniconda3/envs/bkdev/lib/python3.9/json/encoder.py in iterencode(self, o, _one_shot)
    255                 self.key_separator, self.item_separator, self.sort_keys,
    256                 self.skipkeys, _one_shot)
--> 257         return _iterencode(o, 0)
    258 
    259 def _make_iterencode(markers, _default, _encoder, _indent, _floatstr,

~/projects/bokeh/bokeh/core/json_encoder.py in default(self, obj)
    259             return obj.to_css()
    260         else:
--> 261             return self.transform_python_types(obj)
    262 
    263 #-----------------------------------------------------------------------------

~/projects/bokeh/bokeh/core/json_encoder.py in transform_python_types(self, obj)
    228 
    229         else:
--> 230             return super().default(obj)
    231 
    232     def default(self, obj: Any) -> Any:

~/miniconda3/envs/bkdev/lib/python3.9/json/encoder.py in default(self, o)
    177 
    178         """
--> 179         raise TypeError(f'Object of type {o.__class__.__name__} '
    180                         f'is not JSON serializable')
    181 

TypeError: Object of type Image is not JSON serializable
```

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by orena1**
See https://github.com/bokeh/bokeh/pull/12775#issuecomment-1435162651
Basically to fix this I had to do the following:
Create a list that keeps which values pass validation and run transform on the first param that pass the validation, but there are some edge cases:
- if all params fail validation, just run transform 1-by-1
- validation can be disabled by the user, in the case again just run transform 1-by-1
- sometimes when all params fail validation, convertors are executed here: https://github.com/bokeh/bokeh/blob/226fb6aff189458dd5e5616c8fc375494b05ec2b/src/bokeh/core/property/bases.py#L356, but this is via is_valid which does not update the valids list in Either, so once we find a valid converter and value, we need to re-run the validation function to update the valids list.

- [x] issues: fixes #11830

### CODE REVIEW

**Comment by orena1** [comment]
I see that some tests are failing, but I think they should not, for example:
https://github.com/bokeh/bokeh/blob/226fb6aff189458dd5e5616c8fc375494b05ec2b/tests/unit/bokeh/core/property/test_dataspec.py#L168

The input is a tuple and not a string, this is the order of the Either:
`[String, Nullable(Color), Instance(Value), Instance(Field), Instance(Expr), Struct(value=Nullable(Color), transform=Instance(Transform)), Struct(field=String, transform=Instance(Transform)), Struct(expr=Instance(Expression), transform=Instance(Transform))]`

Therefore the validations list is:
`[False, True, False, False, False, False, False, False]`

So the transform of `Nullable(Color)` should be executed and not the one of string, after my pull request this is what happens, and the output is `rgb(128, 255, 124)` but the test expects `(128, 255, 124)`

What are your thoughts? @bryevdv 

Thanks,
Oren

File: src/bokeh/core/property/either.py [hunk 1]
```diff
@@ -81,20 +81,24 @@ def __init__(self, type_param0: TypeOrInst[Property[Any]], *type_params: TypeOrI
...
             except ValueError:
                 pass
 
         raise ValueError(f"Could not transform {value!r}")
 
     def validate(self, value: Any, detail: bool = True) -> None:
         super().validate(value, detail)
-
-        if any(param.is_valid(value) for param in self.type_params):
+        self.valids = [param.is_valid(value) for param in self.type_params]
```
Comment (bryevdv) [inline src/bokeh/core/property/either.py:100]: I think if we want to efficiently store the first valid param, without excess computation, we could use this:
```python
self.valid = next((p for for p in self.type_params if p.is_valid(value)), None)
```
if there are no valid params then the value of `self.valid` will be None

File: src/bokeh/core/property/either.py [hunk 2]
```diff
@@ -81,20 +81,24 @@ def __init__(self, type_param0: TypeOrInst[Property[Any]], *type_params: TypeOrI
         super().__init__(type_param0, *type_params, default=default, help=help)
         for tp in self.type_params:
             self.alternatives.extend(tp.alternatives)
+        self.valids = []
```
Comment (bryevdv) [inline src/bokeh/core/property/either.py:84]: If we make the change below, then 
```suggestion
        self.valid = None
```

File: src/bokeh/core/property/either.py [hunk 3]
```diff
@@ -81,20 +81,24 @@ def __init__(self, type_param0: TypeOrInst[Property[Any]], *type_params: TypeOrI
...
-        for param in self.type_params:
+        for i, param in enumerate(self.type_params):
             try:
-                return param.transform(value)
+                # if valid was not executed, or all valids failed - transform one by one
+                # else: run transform on the first param that pass validation
+                if len(self.valids)==0 or not any(self.valids) or self.valids[i]: 
+                    return param.transform(value)
             except ValueError:
                 pass
```
Comment (bryevdv) [inline src/bokeh/core/property/either.py:95]: If we make the change in for `self.valid` then maybe like:
```python
if self.valid: 
    return self.valid.transform(value)

for param in self.type_params:
    try:
        return param.transform(value)
    except ValueError:
        pass
```
This avoids the loop if there is a valid type (and avoids making the loop complicated to try to handle all cases itself)

Question: if `self.valid` is not None, can `self.valid.transform` fail? If so we might `try...except: pass` around the first conditional, to let it fall through to the loop in that case (or else raise an explicit informative error if that situation is not reasonable)

**Review by bryevdv — COMMENTED** [review]
Hi @orena1 thanks for you patience, the last couple of months have been challenging for me. I took a closer look and had some suggestions. Please let me know what you think

**Comment by bryevdv** [comment]
Hi @orena1 do you expect to be able to continue work on this PR and the comments above?

**Comment by orena1** [comment]
Hi @bryevdv, thanks for the reminder,  I've just answered to your main comment (the rest I think are a result of that one)..

File: src/bokeh/core/property/either.py (hunk 1 above)
Comment (orena1) [reply src/bokeh/core/property/either.py:100]: Hi, thanks for the comment, but sometime it can happen that 2 pass validation, and we will try to run first the first one that pass validation, and if it does not work, the second one that pass validation.
e.g.
I can sometime see that `self.valid=[False, False, True, False, True]`
So in the case, we want to try to run the 3rd transform, and if it does not work the last transform.
What do you think?

Comment (bryevdv) [reply src/bokeh/core/property/either.py:100]: I think that if the the third option returns valid, but then subsequently fails to transform, that there must be a bug somewhere else. A passing validation check is supposed to mean exactly that the property _will_ accept the value. If it does not, something else is seriously wrong. Can you provide a concrete test case to demonstrate the situation you describe?

Comment (orena1) [reply src/bokeh/core/property/either.py:100]: Yeah, I need to find the edge case again, i'll get back to you. It also relates to your question below:
"Question: if self.valid is not None, can self.valid.transform fail? If so we might try...except: pass around the first conditional,  ..."

Comment (bryevdv) [reply src/bokeh/core/property/either.py:100]: cc @mattpap do you have any insights to add? If a property validates a value, I think that a subsequent `transform` should be considered expected to work. To answer my own question, I think if the transform fails with `self.valid` then an informative error should be raised.

**Comment by bryevdv** [comment]
@mattpap unless you have any thoughts on this, I am going to close. I do think we need to find a solution that short-circuits and avoids storing as much state

**Comment by bryevdv** [comment]
I am going to close this for lack of activity and some unresolved questions

**Comment by orena1** [comment]
Yeah, sorry, did not find the bandwidth to get back to it.
